### PR TITLE
replace asserts with nicer ValueErrors in gyb

### PR DESCRIPTION
**Pull request by JHoloboski**
Saw the fixme in this file on the assert, so:
- Raise ValueErrors where asserts used to be, giving the user some feedback

### CODE REVIEW

**Comment by gribozavr** [comment]
Thanks!  Please clean up the patch so that there are no unrelated whitespace changes.

**Comment by JHoloboski** [comment]
@gribozavr Done!

File: utils/gyb.py [hunk 1]
```diff
@@ -464,10 +464,14 @@ def tokenGenerator(self, baseTokens):
                 if (kind == 'gybBlockOpen'):
                     # Absorb any '}% <optional-comment> \n'
                     m2 = gybBlockClose.match(self.template, closePos)
-                    assert m2, "Invalid block closure" # FIXME: need proper error handling here.
+                    if m2 is None:
```
Comment (gribozavr) [inline utils/gyb.py]: Is this equivalent?  I thought the equivalent would be `if not m2:`.

File: utils/gyb.py [hunk 2]
```diff
@@ -464,10 +464,14 @@ def tokenGenerator(self, baseTokens):
...
+                        raise ValueError("Invalid block closure")
                     nextPos = m2.end(0)
                 else:
-                    assert kind == 'substitutionOpen'
```
Comment (gribozavr) [inline utils/gyb.py]: This assert is just an assert that would catch a logic error.  Please revert this part.

File: utils/gyb.py (hunk 1 above)
Comment (gribozavr) [inline utils/gyb.py]: It would be also great to add a test for this.

Comment (JHoloboski) [inline utils/gyb.py]: Yeah that would be better, fixing.

File: utils/gyb.py (hunk 2 above)
Comment (JHoloboski) [inline utils/gyb.py]: Gotcha, done.

File: utils/gyb.py [hunk 3]
```diff
@@ -655,7 +659,14 @@ def execute(self, context):
         # Execute the code with our __children__ in scope 
         context.localBindings['__children__'] = self.children
         result = eval(self.code, context.localBindings)
-        assert context.localBindings['__children__'] is self.children
+
+        if context.localBindings['__children__'] is not self.children:
+            raise ValueError(
+                "Expected bindings to be {}, got {}".format(
```
Comment (gribozavr) [inline utils/gyb.py]: This error message isn't precise enough.  It should say, I think, "The code is not allowed to mutate __children__".

Here's a test:
```
%{
__children__ = 42
}%
```

Comment (JHoloboski) [inline utils/gyb.py]: Sure, that makes sense.

File: utils/gyb.py [hunk 4]
```diff
@@ -655,7 +656,9 @@ def execute(self, context):
         # Execute the code with our __children__ in scope 
         context.localBindings['__children__'] = self.children
         result = eval(self.code, context.localBindings)
-        assert context.localBindings['__children__'] is self.children
+
+        if context.localBindings['__children__'] is not self.children:
+            raise ValueError("The code is not allowed to mutate children")
```
Comment (gribozavr) [inline utils/gyb.py]: This should say `__children__`, with underscores, as it refers to the actual variable name.

Could you add tests that I mentioned in the other messages?  gyb has a `--test` mode.

Comment (JHoloboski) [inline utils/gyb.py]: Looking, bit new to doctest :sweat_smile:

Comment (JHoloboski) [inline utils/gyb.py]: Could you be more specific on how you'd like me to test those two functions?

Comment (gribozavr) [inline utils/gyb.py]: You can check that executing the source, for example:
```
%{
__children__ = 42
}%
```
produces the desired error message.

Comment (JHoloboski) [inline utils/gyb.py]: Running the test fails on the master branch.

```bash
./gyb --test gyb.py
```
Unless I'm running the tests incorrectly? How can I make the tests pass on master so I can add my tests?

**Comment by gribozavr** [comment]
@JHoloboski Thank you!  Would you mind squashing 5 commits into one and rebasing on top of current master?

**Comment by JHoloboski** [comment]
@gribozavr Done :)

**Comment by dabrahams** [comment]
Did anyone look at the actual results of this?  What does the "nicer feedback" look like?

**Pull request merged**